kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/testplugins/SaslPlainInitiationFilter.java

```diff
@@ -0,0 +1,155 @@
+        return new RequestHeaderData()
+                .setRequestApiKey(apiKey.id)
+                .setRequestApiVersion(apiVersion)
+                .setClientId("sasl-initiator");
+    }
+
+    @Override
+    public CompletionStage<ResponseFilterResult> onApiVersionsResponse(short apiVersion,
+                                                                       ResponseHeaderData header,
+                                                                       ApiVersionsResponseData apiVersionsResponse,
+                                                                       FilterContext context) {
+        short handshakeMaxVersion = apiVersionsResponse.apiKeys().find(ApiKeys.SASL_HANDSHAKE.id).maxVersion();
+        short authenticateMaxVersion = apiVersionsResponse.apiKeys().find(ApiKeys.SASL_AUTHENTICATE.id).maxVersion();
+        LOGGER.info("Received API_VERSIONS response, maxHandshakeVersion: {}, maxAuthenticateVersion: {}", handshakeMaxVersion, authenticateMaxVersion);
+        this.saslHandshakeVersion = (short) Math.max(handshakeMaxVersion, 1);
+        this.saslAuthenticateVersion = (short) Math.max(authenticateMaxVersion, 2);
+        LOGGER.info("Will use handshakeVersion: {}, authenticateVersion: {}", this.saslHandshakeVersion, this.saslAuthenticateVersion);
+
+        if (state == State.AWAITING_API_VERSIONS) {
+            LOGGER.info("Sending SASL_HANDSHAKE request");
+            state = State.AWAITING_SASL_HANDSHAKE;
+            return context.<SaslHandshakeResponseData> sendRequest(
+                    requestHeader(ApiKeys.SASL_HANDSHAKE, saslHandshakeVersion),
+                    new SaslHandshakeRequestData()
+                            .setMechanism("PLAIN"))
+                    .thenCompose(handshakeResponse -> {
+                        if (handshakeResponse.errorCode() == Errors.NONE.code()) {
+                            LOGGER.info("Sending SASL_AUTHENTICATE request");
+                            state = State.AWAITING_SASL_AUTHENTICATE;
+                            return context.<SaslAuthenticateResponseData> sendRequest(
+                                    requestHeader(ApiKeys.SASL_AUTHENTICATE, saslAuthenticateVersion),
+                                    new SaslAuthenticateRequestData()
+                                            .setAuthBytes(("\0" + username + "\0" + password).getBytes(StandardCharsets.UTF_8)))
+                                    .thenCompose(authenticateResponse -> {
+                                        if (authenticateResponse.errorCode() == Errors.NONE.code()) {
+                                            for (BufferedRequest bufferedRequest : bufferedRequests) {
+                                                LOGGER.info("Forwarding buffered {} request", ApiKeys.forId(bufferedRequest.header.apiKey()));
+                                                context.forwardRequest(bufferedRequest.header, bufferedRequest.request);
```

Comment:
you want to complete the BufferedRequest.future so that the core will start reading the client again.
I realise this is only test/example code, but you should also clear the bufferedRequst.list.  Test/examole code has a habit of finding its way into prod :).

Reply:
I've "fixed" this one, but I don't think it's quite correct (although the test passes). This filter (and the others!) really needs a unit test which drives all the possible requests from the client, but I think that's out of scope for this PR which is really about the API, not these filters.